### PR TITLE
sql: improve statisticsMutator

### LINKED ISSUE #62133: roachtest: sqlsmith invalid interface conversion from DBytes to DGeometry
[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2785193&tab=buildLog) on [release-21.1@ba54410af79c0481e1e8f8fc45a9b132ee09a2b4](https://github.com/cockroachdb/cockroach/commits/ba54410af79c0481e1e8f8fc45a9b132ee09a2b4):

```
The test failed on branch=release-21.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: interface conversion: tree.Datum is *tree.DBytes, not *tree.DGeometry
		stmt:
		WITH
			with_238 (col_1444) AS (SELECT * FROM (VALUES ('f':::rand_typ_2)) AS tab_631 (col_1444)),
			with_239 (col_1445, col_1446)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									'-13 years -8 mons -572 days -12:56:55.072675':::INTERVAL,
									COALESCE(
										((('1987-10-03':::DATE::DATE || NULL::DATE[])::DATE[]::DATE[] || ARRAY['1980-03-13':::DATE,'1989-01-13':::DATE,'1982-02-27':::DATE]::DATE[])::DATE[]::DATE[] || ARRAY['1970-08-22':::DATE,'1974-01-12':::DATE]::DATE[])::DATE[],
										ARRAY['1981-10-15':::DATE,'1985-10-11':::DATE]
									)
								)
						)
							AS tab_632 (col_1445, col_1446)
				)
		SELECT
			count(*) AS col_1447
		FROM
			defaultdb.public.table1@[0] AS tab_633
		WHERE
			true
		GROUP BY
			tab_633.col1_0
		HAVING
			st_containsproperly(tab_633.col1_0::GEOMETRY, '0103000000010000000400000080243A9AF5B79F41144F37B249FBF541F09535D6F9A9CB414C8DB4EEA99BF541C678E01C19C1F1C174584659C4D1FD4180243A9AF5B79F41144F37B249FBF541':::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			53:::INT8;

```

<details><summary>More</summary><p>

Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2785193&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

### PR DESCRIPTION
**Pull request by mgartner**
#### sql: refactor statisticsMutator

Release note: None

#### sql: generate random inverted index histograms in statisticsMutator

Fixes #62133

Release note: None

### CODE REVIEW

File: pkg/sql/mutations/mutations.go [hunk 1]
```diff
@@ -300,6 +256,8 @@ func statisticsMutator(
 				allStats = append(allStats, cs)
 			}
 			b, err := json.Marshal(allStats)
+			fmt.Println("MARSHALLED")
```
Comment (otan) [inline pkg/sql/mutations/mutations.go:259]: is this intended?

Comment (mgartner) [reply pkg/sql/mutations/mutations.go]: damnit, no. fixed.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MW1BGHJ2myrm68IOLJq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @otan, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MW1BArt4J-253vmG98S:-MW1BArt4J-253vmG98T:bgloiqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L212)):*
> ```Go
> 			}
> 			colType := tree.MustBeStaticallyKnownType(col.Type)
> 			h := randHistogram(rng, colType)
> ```

Do we maybe want to not add the histogram with some random chance?
___
*[pkg/sql/mutations/mutations.go, line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MW1B5DV64hNjtSzchIS:-MW1B5DV64hNjtSzchIT:bpgwnpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L277)):*
> ```Go
> 
> // randHistogram generates a histogram for the given type with random histogram
> // buckets.
> ```

Explain that if the colType is JSON, we create an inverted index histogram.

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @otan)

___
*[pkg/sql/mutations/mutations.go, line 298 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWQoVQ7B7uWRQwcduHM:-MWQoVQ7B7uWRQwcduHN:btpoqpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L298)):*
> ```Go
> 			var da rowenc.DatumAlloc
> 			for i := range encs {
> 				enc, err := rowenc.EncodeTableKey(nil, da.NewDBytes(tree.DBytes(encs[i])), encoding.Ascending)
> ```

are you sure that you need to call `EncodeTableKey` here?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MW1BArt4J-253vmG98S:-MWVWemw7KK6m13FBOYN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L212)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we maybe want to not add the histogram with some random chance?
</blockquote></details>

Done.
___
*[pkg/sql/mutations/mutations.go, line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MW1B5DV64hNjtSzchIS:-MWVVJ8cAY1Cj3AlPwqR:b-idw4kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L277)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Explain that if the colType is JSON, we create an inverted index histogram.
</blockquote></details>

I don't know that a histogram is specifically an "inverted index histogram", but I explained the byte encoding a bit for JSON types.
___
*[pkg/sql/mutations/mutations.go, line 298 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWQoVQ7B7uWRQwcduHM:-MWVVwaE3ssv8_nh2lOo:b-zhjsb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d77d215e84698cad10d176ebc6ea4ee2c163f81b/pkg/sql/mutations/mutations.go#L298)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

are you sure that you need to call `EncodeTableKey` here?
</blockquote></details>

Yes. I added a comment to explain. `JSONStatistic.SetHistogram` is called with the histogram we generate here and it tries to decode the upper-bound. Therefore we need to encode the inverted index key as a bytes datum here, otherwise decoding fails.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MWW7EiuEEDORpdrD2Hv:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @RaduBerinde)

**Review by rytaft — COMMENTED** [review]
Would be great to get this in (including a couple of changes I suggested below, since I think it will fix #62133)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MWfEE8l36dvDYJU3k2Z:b5q6qs7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 285 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE-CG4U6wMzI0UmGy:-MWfE-CG4U6wMzI0UmGz:b-awe3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L285)):*
> ```Go
> func randHistogram(rng *rand.Rand, colType *types.T) stats.HistogramData {
> 	histogramColType := colType
> 	if histogramColType.Family() == types.JsonFamily {
> ```

Can you change this to check if the type is invertedIndexable?
___
*[pkg/sql/mutations/mutations.go, line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE40M9hCQWvjnzVq7:-MWfE40M9hCQWvjnzVq8:b-awe3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L296)):*
> ```Go
> 	for i, numDatums := 0, rng.Intn(10); i < numDatums; i++ {
> 		upper := rowenc.RandDatum(rng, colType, false /* nullOk */)
> 		if colType.Family() == types.JsonFamily {
> ```

Can you change this to check if the type is invertedIndexable?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 285 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE-CG4U6wMzI0UmGy:-MWg0jtb6fuQIoCiA8oD:bi4cos3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L285)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you change this to check if the type is invertedIndexable?
</blockquote></details>

Ya this actually confused me a bit. `rowenc.EncodeTableKey` can encode a DArray, but I suppose this is not how an inverted index on an ARRAY is actually encoded?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 285 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE-CG4U6wMzI0UmGy:-MWg2aAUAWvwjmpRwmJD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L285)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Ya this actually confused me a bit. `rowenc.EncodeTableKey` can encode a DArray, but I suppose this is not how an inverted index on an ARRAY is actually encoded?
</blockquote></details>

Done.
___
*[pkg/sql/mutations/mutations.go, line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE40M9hCQWvjnzVq7:-MWg2_5XC_dU_VoqKqge:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L296)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you change this to check if the type is invertedIndexable?
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 2 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MWg4fxM-mkYhTVC7MYY:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @RaduBerinde)

___
*[pkg/sql/mutations/mutations.go, line 285 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWfE-CG4U6wMzI0UmGy:-MWg4Dz1C06htzunKrRx:b-m6wt0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e970d025310a1c5242dbd81c99edc76af001b8f5/pkg/sql/mutations/mutations.go#L285)):*
> rowenc.EncodeTableKey can encode a DArray, but I suppose this is not how an inverted index on an ARRAY is actually encoded?

I believe that for a short period of time we supported both forward and inverted indexes on arrays, but we removed support for forward indexes on arrays since it made things complicated. Maybe we'll add it back in the future, but for now I don't think you need to worry about it.

**Comment by mgartner** [comment]
This is still broken. Getting `panic: trying to apply inverted index to unsupported type geometry` when running `TestSetups`.

**Review by rytaft — COMMENTED** [review]
You'll need to call `EncodeGeoInvertedIndexTableKeys` for the spatial types. You can probably just make a dummy index descriptor with the `GeoConfig` set to `DefaultGeographyIndexConfig` or `DefaultGeometryIndexConfig`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MWjIm01739NUCdHHKEn:b-tqlqyb)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @otan, and @RaduBerinde)

___
*[pkg/sql/mutations/mutations.go, line 297 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWjKVF1BPyXQOMFYC0Z:-MWjJ-6k3xaN30F0TODu:b-5klnto) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f5b61a5ff041d025d5c5ff4d6adac6fa2e4772e/pkg/sql/mutations/mutations.go#L297)):*
> ```Go
> 		upper := rowenc.RandDatum(rng, colType, false /* nullOk */)
> 		if colinfo.ColumnTypeIsInvertedIndexable(colType) {
> 			encs, err := rowenc.EncodeInvertedIndexTableKeys(upper, nil, descpb.EmptyArraysInInvertedIndexesVersion)
> ```

This is where you'll need to call `EncodeGeoInvertedIndexTableKeys` if you have a spatial type.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @otan, @RaduBerinde, and @rytaft)

___
*[pkg/sql/mutations/mutations.go, line 297 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-MWjKVF1BPyXQOMFYC0Z:-MWjos3cA9WTjESZsta0:bpaetst) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f5b61a5ff041d025d5c5ff4d6adac6fa2e4772e/pkg/sql/mutations/mutations.go#L297)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This is where you'll need to call `EncodeGeoInvertedIndexTableKeys` if you have a spatial type.
</blockquote></details>

Thanks for the tips! Done.

**Review by rytaft — APPROVED** [review]
Looks like you still have a test failure... but :lgtm: otherwise

> Reviewed 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62173#-:-MWk7PM97v4fUn-epL30:b-8164r1)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @RaduBerinde)

**Comment by mgartner** [comment]
TFTRs!

bors r+

**Pull request merged**